pkg/kubelet/kubelet.go

```diff
@@ -929,34 +929,107 @@ type Kubelet struct {
+
+	rootDirectory string
 
 	lastObservedNodeAddressesMux sync.RWMutex
 	lastObservedNodeAddresses    []v1.NodeAddress
 
 	// onRepeatedHeartbeatFailure is called when a heartbeat operation fails more than once. optional.
 	onRepeatedHeartbeatFailure func()
 
-	// podWorkers handle syncing Pods in response to events.
+	// podManager stores the desired set of admitted pods and mirror pods that the kubelet should be
+	// running. The actual set of running pods is stored on the podWorkers. The manager is populated
+	// by the kubelet config loops and may be consulted by other components that need to see the set
+	// of desired pods. Note that not all desired pods are running, and not all running pods are in
+	// the podManager - for instance, force deleting a pod from the apiserver will remove it from
+	// the podManager, but the pod may still be terminating and tracked by the podWorkers. Components
+	// that need to know the actual consumed resources of the node or are driven by podWorkers and
+	// the sync*Pod methods	(status, volume, stats) should also consult the podWorkers when
+	// reconciling.
+	//
+	// TODO: review all kubelet components that need the actual set of pods (vs the desired set)
+	// and update them to use podWorkers instead of podManager. This may introduce latency in some
+	// methods, but avoids race conditions and correctly accounts for terminating pods that have
+	// been force deleted or static pods that have been updated.
+	podManager kubepod.Manager
+
+	// podWorkers is responsible for driving the lifecycle state machine of each pod. The worker is
+	// notified of config changes, updates, periodic reconciliation, container runtime updates, and
+	// evictions of all desired pods and will invoke reconciliation methods per pod in separate
+	// goroutines. The podWorkers are authoritative in the kubelet for what pods are actually being
+	// run and their current state:
+	//
+	// * syncing: pod should be running (syncPod)
+	// * terminating: pod should be stopped (syncTerminatingPod)
+	// * terminated: pod should have all resources cleaned up (syncTerminatedPod)
+	//
+	// and invoke the handler methods that correspond to each state. Components within the
+	// kubelet that need to know the phase of the pod in order to correctly set up or tear down
+	// resources must consult the podWorkers.
+	//
+	// Once a pod has been accepted by the pod workers, no other pod with that same UID (and
+	// fullName, for static pods) will be started until the first pod has fully terminated and
+	// been cleaned up by SyncKnownPods. This means a pod may be desired (in API), admitted (in
+	// pod manager), and requested (by invoking UpdatePod) but not start for an arbitrarily long
+	// interval because a prior pod is still terminating.
+	//
+	// As an event-driven (by UpdatePod) controller, the podWorkers must periodically be resynced
+	// by the kubelet invoking SyncKnownPods with the desired state (admitted pods in podManager).
```

Comment:
nit: It might be also worth mentioning that kubelet is responsible for calling `HandlePodCleanups` periodically during house keeping (which calls `SyncKnownPods` underneath), which includes terminating pod workers and unwanted pods, e.g. those that were force killed.

Reply:
good point